wai-extra/Network/Wai/Middleware/RequestSizeLimit.hs

```diff
@@ -0,0 +1,69 @@
+module Network.Wai.Middleware.RequestSizeLimit (requestSizeLimit) where
+
+import Network.Wai
+import Network.Wai.Request
+import qualified Data.ByteString.Char8 as S8
+import qualified Data.ByteString.Lazy as BSL
+import           Data.Word                    (Word64)
+import           Data.IORef              (newIORef, readIORef, writeIORef)
+import Network.HTTP.Types.Status (requestEntityTooLarge413)
+import qualified Data.ByteString.Lazy.Char8 as LS8
+
+requestSizeLimit :: Word64 -> Middleware
+requestSizeLimit maxLen app req sendResponse = do
+  putStrLn "In request size limit"
+  case requestBodyLength req of
+    KnownLength actualLen -> if actualLen > maxLen then
+      sendResponse (tooLargeResponse maxLen actualLen)
+      else app req sendResponse
+    ChunkedBody -> do
+        ref <- newIORef maxLen
+        let newReq = req
+                { requestBody = do
+                    bs <- getRequestBodyChunk req
+                    remaining <- readIORef ref
+                    let len = fromIntegral $ S8.length bs
+                        remaining' = remaining - len
+                    if remaining < len
+                        then do
+                            putStrLn "Too large; sending too large response here"
+                            -- This won't work, because execution continues after sendResponse
+                            sendResponse $ tooLargeResponse maxLen len
+                            putStrLn "Too large; just sent too large response here"
+                            pure ""
```

Comment:
Ok, I've been pulling in the code from the Yesod implementation, but this part is a little tricky. In Yesod, it does `throwIO $ HCWai $ tooLargeResponse maxLen len`, which Yesod catches and interprets as "send this WAI response".

I initially tried just using `sendResponse` here, but that doesn't work because execution continues past this point. Should I try a similar pattern as Yesod, where I throw an exception here, and attempt to catch it and give the 413? I'm not sure if WAI has this kind of pattern

Just worked on this a little bit tonight, haven't investigated too deeply so need to do more research, but if you have an idea off-hand I'd take that

Reply:
```
newtype TooLargeRequestException = TooLargeRequestException Response

instance Show TooLargeRequestException where
    show (TooLargeRequestException r) = "TooLargeRequestException: The request exceeded the total available limit. This exception is automatically caught by WAI to provide a proper response"
instance Exception TooLargeRequestException

requestSizeLimit :: Word64 -> Middleware
requestSizeLimit maxLen app req sendResponse = do
  putStrLn "In request size limit"
  case requestBodyLength req of
    KnownLength actualLen -> if actualLen > maxLen then
      sendResponse (tooLargeResponse maxLen actualLen)
      else app req sendResponse
    ChunkedBody -> do
        ref <- newIORef maxLen
        let newReq = req
                { requestBody = do
                    bs <- getRequestBodyChunk req
                    remaining <- readIORef ref
                    let len = fromIntegral $ S8.length bs
                        remaining' = remaining - len
                    if remaining < len
                        then do
                            putStrLn "Too large; sending too large response here"
                            throwIO (TooLargeRequestException $ tooLargeResponse maxLen len)
                            -- This won't work, because execution continues after sendResponse
                            sendResponse $ tooLargeResponse maxLen len
                            putStrLn "Too large; just sent too large response here"
                            pure ""
                        else do
                            writeIORef ref remaining'
                            return bs
                }
        app newReq sendResponse `catch` \(TooLargeRequestException resp) -> sendResponse resp
```

Ok, this seems to work. Is this pattern acceptable in WAI?

Reply:
That's great! One request though: please take out the `putStrLn`, middlewares shouldn't generate output like that in general. The exception being thrown should be sufficient. If someone wants to generate a message, they can `onException` it. Alternatively: you could take a callback function to perform an action before throwing the exception.

Also, what's the point of the `sendResponse` and other two lines after the `throwIO` call?

Reply:
Oh yeah, this isn’t the final code, this is the middle of me debugging stuff. I just wanted to check the exception approach was correct.